tests/testsuite/binary_name.rs

```diff
@@ -0,0 +1,157 @@
+use cargo_test_support::project;
+
+#[cargo_test]
+// This test checks if Cargo produces a binary with the name we provide in `Cargo.toml` in the
+// `filename` parameter under the `[[bin]]` section.
+fn simple_binary_name() {
+    // Create the project.
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                cargo-features = ["different-binary-name"]
+
+                [project]
+                name =  "foo"
+                version = "0.0.1"
+                authors = []
+
+                [[bin]]
+                name = "foo"
+                filename = "007bar"
+                path = "src/main.rs"
+            "#,
+        )
+        .file("src/main.rs", "fn main() { assert!(true) }")
+        .build();
+
+    // Run cargo build.
+    p.cargo("build").masquerade_as_nightly_cargo().run();
+
+    // Check the name of the binary that cargo has generated.
+    // A binary with the name of the crate should NOT be created.
+    let foo_path = p.bin("foo");
+    assert!(!foo_path.is_file());
+    // A binary with the name provided in `filename` parameter should be created.
+    let bar_path = p.bin("007bar");
+    assert!(bar_path.is_file());
+}
+
+#[cargo_test]
+// Same test as `simple_binary_name` but without providing `path` in the `[[bin]]` section.
+fn simple_binary_name2() {
+    // Create the project.
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                cargo-features = ["different-binary-name"]
+
+                [project]
+                name =  "foo"
+                version = "0.0.1"
+                authors = []
+
+                [[bin]]
+                name = "foo"
+                filename = "007bar"
+            "#,
+        )
+        .file("src/main.rs", "fn main() { assert!(true) }")
+        .build();
+
+    // Run cargo build.
+    p.cargo("build").masquerade_as_nightly_cargo().run();
+
+    // Check the name of the binary that cargo has generated.
+    // A binary with the name of the crate should NOT be created.
+    let foo_path = p.bin("foo");
+    assert!(!foo_path.is_file());
+    // A binary with the name provided in `filename` parameter should be created.
+    let bar_path = p.bin("007bar");
+    assert!(bar_path.is_file());
+}
+
+// Check if deps info, is populated with the correct dependencies.
+#[cargo_test]
+fn check_deps_content() {
```

Comment:
Can this test be merged with `simple_binary_name`?  Each test has a small cost, and we don't want to add them too extravagantly.

Reply:
So, basically, I should create new tests only if the content of `cargo.toml` changes. Is that correct? We want to perform as many checks as we can in the same test method (within good reason) as long as all of them can use the same `cargo.toml` content.

Reply:
Unfortunately there aren't any specific rules to follow.  It's more of striking a balance.  You don't want a single test to be too long, and cover too much.  That can make it hard to understand, and if you are trying to debug just one part of it, all the extraneous stuff gets in the way.  On the other hand, having too many small tests makes the tests take longer to run.

But yea, my general rule is if `Cargo.toml` needs to be different, then it can go in a separate test.

It's mostly a matter of taste, and different people will have different opinions on how to balance it.

Reply:
Hm, I thought I responded to this, but I think my comment got eaten. There aren't any specific rules. But yea, when different steps of a test have the same `Cargo.toml`, I usually place them together in the same test. There is a balance between having too many individual tests (which slows things down), and having too many steps in a single test (which can make the test hard to work on and debug). Different people will have different opinions about exactly how to structure things, and it is mostly just a judgement call.